### PR TITLE
Implement better user password validation

### PR DESCRIPTION
**Pull request by zoldar**
[password_strength_demo2.webm](https://github.com/plausible/analytics/assets/588351/92662fc3-bc06-4bb4-9d15-94b2e5b6624b)

### Changes

- Modify length requirements - password must be minimum 12 characters long now with upper bound of 128
- Move registration form to LiveView, implementing instant feedback as form is filled
- Implement password strength validation with a detailed visual and human readable feedback whenever possible
- Deduplicate logic from registration via invitation flow
- Revise password reset and set forms, implementing LV-base password validation feedback
- Make form more password manager friendly via `autocomplete` (and in some cases `id`) property 

### TODO before final review/merge

- [x] re-evaluate forms in dark mode
- [x] ~decide whether to expose set password form as it's currently not linked anywhere (may be done in a follow-up)~ `/password` is a leftover from legacy flow and will be removed completely
- [x] deploy to staging and conduct more live testing
- [x] ~consider validating email format server-side, instead of using "email" input type (may be done in a follow-up)~ (might be subject to change in a follow-up)

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [x] Entry has been added to changelog
- [ ] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [x] The UI has been tested both in dark and light mode
- [ ] This PR does not change the UI

### CODE REVIEW

File: lib/plausible_web/live/set_password_form.ex [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  import PlausibleWeb.Live.Components.Form
+
+  alias Plausible.Auth
+  alias Plausible.Repo
+
+  def mount(_params, %{"reset_token" => reset_token}, socket) do
+    # by that point token should be already verified
+    {:ok, %{email: email}} = Auth.Token.verify_password_reset(reset_token)
+    user = Repo.get_by!(Auth.User, email: email)
+    changeset = Auth.User.changeset(user)
```
Comment (aerosol) [inline lib/plausible_web/live/set_password_form.ex:18]: ditto, `assign_new` will let you share the assigns across both `mount` calls

**Review by aerosol — COMMENTED** [review]
Comments inline.

When entering spaces into the password field, we're getting two error messages but one should be enough.

image

File: lib/plausible_web/templates/auth/password_form.html.eex [hunk 2]
```diff
@@ -1,14 +1,5 @@
...
-  </div>
-  <%= submit "Set password →", class: "button mt-4 w-full" %>
-  <p class="text-center text-gray-500 text-xs mt-4">
-    Don't have an account? <%= link("Register", to: "/register", class: "underline text-gray-800 dark:text-gray-200") %> instead.
-  </p>
-<% end %>
+<%= live_render(@conn, PlausibleWeb.Live.SetPasswordForm,
+    container: {:div, class: "contents"},
+    session: %{},
+    router: PlausibleWeb.Router
```
Comment (aerosol) [inline lib/plausible_web/templates/auth/password_form.html.eex:4]: I don't think that does anything, we should probably remove from other places too.

File: lib/plausible/auth/user.ex [hunk 3]
```diff
@@ -88,6 +90,47 @@ defmodule Plausible.Auth.User do
...
   end
 
+  def password_strength(changeset) do
+    if password = get_field(changeset, :password) do
+      existing_phrases =
+        []
+        |> maybe_add_phrase(get_field(changeset, :name))
+        |> maybe_add_phrase(get_field(changeset, :email))
+
+      # checking only first 32 characters of the password
```
Comment (aerosol) [inline lib/plausible/auth/user.ex:100]: Minor. It's an extreme case but perhaps with longer passwords we should just accept them. The feedback for `(32 x "a") + some strong password here` is incorrect then.

File: lib/plausible_web/live/register_form.ex [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    ~H"""
+    <.form
+      :let={f}
+      for={@form}
+      phx-change="validate"
+      phx-submit="register"
+      phx-trigger-action={@trigger_submit}
+      class="w-full max-w-md mx-auto bg-white dark:bg-gray-800 shadow-md rounded px-8 py-6 mb-4 mt-8"
+    >
+      <input name="_csrf_token" type="hidden" value={Plug.CSRFProtection.get_csrf_token()} />
```
[the remark this answers is not in the document]
Comment (zoldar) [reply lib/plausible_web/live/register_form.ex:44]: No, it's the same. The problem is, this form is eventually submitted to a standard HTTP endpoint (via `phx-trigger-action` toggle) and it's lacking `_csrf_token` value in the POST payload (it's not necessary when sending data via WS).

File: lib/plausible_web/templates/auth/password_form.html.eex (hunk 2 above)
Comment (zoldar) [reply lib/plausible_web/templates/auth/password_form.html.eex:4]: I'll double check and remove it, thanks 🙌

File: lib/plausible/auth/user.ex (hunk 3 above)
Comment (zoldar) [reply lib/plausible/auth/user.ex:100]: Ah, good point. Yeah, perhaps past that point we should just accept whatever's there 👍

**Comment by zoldar** [comment]
> Comments inline.
> 
> When entering spaces into the password field, we're getting two error messages but one should be enough.
> 
> image

Right, that's because params are trimmed/scrubbed from whitespace at both ends (which I think should still hold, even for passwords). I'll alter the logic so the error gets "deduplicated" in this case as well.

**Comment by aerosol** [comment]
> > Comments inline.
> > When entering spaces into the password field, we're getting two error messages but one should be enough.
> > image
> 
> Right, that's because params are trimmed/scrubbed from whitespace at both ends (which I think should still hold, even for passwords). I'll alter the logic so the error gets "deduplicated" in this case as well.

Cool. I checked if `<space>verystrong<space>` works properly still and it's fine.

**Review by aerosol — COMMENTED** [review]
Just one more thing - if you install the noscript plugin and try to submit the form, it crashes. I believe this is something the bots will try doing, so perhaps we should handle it somehow.

```
Request: POST /register
** (exit) an exception was raised:
    ** (RuntimeError) expected action/2 to return a Plug.Conn, all plugs must receive a connection (conn) and return a connection, got: :user_not_found
        (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:1: PlausibleWeb.AuthController.phoenix_controller_pipeline/2
        (phoenix 1.7.7) lib/phoenix/router.ex:430: Phoenix.Router.__call__/5
        (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.plug_builder_call/2
        (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint."call (overridable 3)"/2
        (plausible 0.0.1) lib/plug/debugger.ex:136: PlausibleWeb.Endpoint."call (overridable 4)"/2
        (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.call/2
        (phoenix 1.7.7) lib/phoenix/endpoint/sync_code_reload_plug.ex:22: Phoenix.Endpoint.SyncCodeReloadPlug.do_call/4
        (plug_cowboy 2.6.1) lib/plug/cowboy/handler.ex:11: Plug.Cowboy.Handler.init/2
        (cowboy 2.10.0) /home/hq1/workspaces/plausible/analytics/deps/cowboy/src/cowboy_handler.erl:37: :cowboy_handler.execute/2
        (cowboy 2.10.0) /home/hq1/workspaces/plausible/analytics/deps/cowboy/src/cowboy_stream_h.erl:306: :cowboy_stream_h.execute/3
        (cowboy 2.10.0) /home/hq1/workspaces/plausible/analytics/deps/cowboy/src/cowboy_stream_h.erl:295: :cowboy_stream_h.request_process/3
        (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```

File: lib/plausible_web/templates/auth/password_form.html.eex (hunk 2 above)
Comment (zoldar) [reply lib/plausible_web/templates/auth/password_form.html.eex]: Done across all `live_render` calls in https://github.com/plausible/analytics/pull/3344/commits/18241f42554db0f6f93bf4a3c277119c0a8c5931

File: lib/plausible_web/live/register_form.ex [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+  import PlausibleWeb.Live.Components.Form
+
+  alias Plausible.Auth
+  alias Plausible.Repo
+
+  def mount(_params, %{"is_selfhost" => is_selfhost} = session, socket) do
+    {changeset, invitation_id} =
+      if invitation_id = session["invitation_id"] do
+        invitation = Repo.get_by!(Auth.Invitation, invitation_id: invitation_id)
```
[the remark this answers is not in the document]
Comment (zoldar) [reply lib/plausible_web/live/register_form.ex]: Done in https://github.com/plausible/analytics/pull/3344/commits/18241f42554db0f6f93bf4a3c277119c0a8c5931 - I've limited it only to expensive, IO/DB-bound assigns, to not obstruct logic flow too much - I think 😅

[the remark this answers is not in the document]
Comment (zoldar) [reply lib/plausible_web/live/register_form.ex]: Ugh, wrong SHA - I meant this one https://github.com/plausible/analytics/pull/3344/commits/3834c5e6579c799a88fbb2094577ef11c0dc6f50

File: lib/plausible_web/live/set_password_form.ex (hunk 1 above)
Comment (zoldar) [reply lib/plausible_web/live/set_password_form.ex]: addressed in https://github.com/plausible/analytics/commit/3834c5e6579c799a88fbb2094577ef11c0dc6f50

File: lib/plausible/auth/user.ex (hunk 3 above)
Comment (zoldar) [reply lib/plausible/auth/user.ex]: Addressed in https://github.com/plausible/analytics/pull/3344/commits/a92e23b9815d0f96536763ab7ba3a6c6e361b510

**Comment by zoldar** [comment]
> > > Comments inline.
> > > When entering spaces into the password field, we're getting two error messages but one should be enough.
> > > image
> > 
> > 
> > Right, that's because params are trimmed/scrubbed from whitespace at both ends (which I think should still hold, even for passwords). I'll alter the logic so the error gets "deduplicated" in this case as well.
> 
> Cool. I checked if `<space>verystrong<space>` works properly still and it's fine.

Fixed it with https://github.com/plausible/analytics/pull/3344/commits/df90f71856e21c4b3c6cbd2cf496d8806a01e75b, finally 😅 

> Just one more thing - if you install the noscript plugin and try to submit the form, it crashes. I believe this is something the bots will try doing, so perhaps we should handle it somehow.
> 
> ```
> Request: POST /register
> ** (exit) an exception was raised:
>     ** (RuntimeError) expected action/2 to return a Plug.Conn, all plugs must receive a connection (conn) and return a connection, got: :user_not_found
>         (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:1: PlausibleWeb.AuthController.phoenix_controller_pipeline/2
>         (phoenix 1.7.7) lib/phoenix/router.ex:430: Phoenix.Router.__call__/5
>         (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.plug_builder_call/2
>         (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint."call (overridable 3)"/2
>         (plausible 0.0.1) lib/plug/debugger.ex:136: PlausibleWeb.Endpoint."call (overridable 4)"/2
>         (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.call/2
>         (phoenix 1.7.7) lib/phoenix/endpoint/sync_code_reload_plug.ex:22: Phoenix.Endpoint.SyncCodeReloadPlug.do_call/4
>         (plug_cowboy 2.6.1) lib/plug/cowboy/handler.ex:11: Plug.Cowboy.Handler.init/2
>         (cowboy 2.10.0) /home/hq1/workspaces/plausible/analytics/deps/cowboy/src/cowboy_handler.erl:37: :cowboy_handler.execute/2
>         (cowboy 2.10.0) /home/hq1/workspaces/plausible/analytics/deps/cowboy/src/cowboy_stream_h.erl:306: :cowboy_stream_h.execute/3
>         (cowboy 2.10.0) /home/hq1/workspaces/plausible/analytics/deps/cowboy/src/cowboy_stream_h.erl:295: :cowboy_stream_h.request_process/3
>         (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
> ```

Yup, so that was an error on my end, not gracefully handling failure modes that shouldn't happen under normal circumstances 😅 Given those endpoints are in plain sight, they definitely should to avoid raising false alarms. I hav done that in https://github.com/plausible/analytics/pull/3344/commits/351a99f56ebc01698157bed2872e3bd221a7cd8b and took the opportunitu to clean up auth code a wee little bit just btw.

**Review by aerosol — APPROVED** [review]
nice work!

File: lib/plausible_web/templates/auth/register_form.html.heex [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      Register your Plausible Analytics account
+    <% else %>
+      Register your 30-day unlimited-use free trial
+    <% end %>
+  </h1>
+  <div class="text-xl font-medium">Set up privacy-friendly analytics with just a few clicks</div>
+</div>
+
+<div class="w-full max-w-3xl mt-4 mx-auto flex flex-shrink-0">
+  <%= live_render(@conn, PlausibleWeb.Live.RegisterForm,
```
Comment (ukutaht) [inline lib/plausible_web/templates/auth/register_form.html.heex:13]: Probably dumb question but why is the live component rendered in a dead view as opposed to using the [LiveRouter](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.Router.html) for the whole template/route?

File: mix.exs [hunk 7]
```diff
@@ -121,7 +121,8 @@ defmodule Plausible.MixProject do
       {:ex_money, "~> 5.12"},
       {:mjml_eex, "~> 0.9.0"},
       {:mjml, "~> 1.5.0"},
-      {:heroicons, "~> 0.5.0"}
+      {:heroicons, "~> 0.5.0"},
+      {:zxcvbn, git: "https://github.com/techgaun/zxcvbn-elixir.git"}
```
Comment (ukutaht) [inline mix.exs:125]: Oh, it's not released on hex.pm?

Besides the recent year regex being outdated, did you find any other shortcomings with this repo?

**Review by ukutaht — APPROVED** [review]

File: lib/plausible_web/templates/auth/register_form.html.heex (hunk 6 above)
Comment (zoldar) [reply lib/plausible_web/templates/auth/register_form.html.heex:13]: Mostly, it's convenience. This way it's easier to use existing facilities: error templates, setting layout, setting plugs within controller consistently between related actions etc. That aside, there's not a whole lot of benefit in LiveRouter routes when we are not using live navigation (yet).

I have made an attempt at moving registration forms over to live routes which resulted in https://github.com/plausible/analytics/pull/3344/commits/d5ff7c43d6473461a2b70773a44da574b0d7220f. It's doable, as can be seen, but I'm going to skip converting reset form for now as this one won't likely benefit from live routes any time soon and requires extra work sorting out rendering error views in live context properly.

Comment (aerosol) [reply lib/plausible_web/templates/auth/register_form.html.heex:13]: Welp, that's a lot of work with for little/no benefit in this case indeed. Good thing we know what it takes now.

File: assets/js/liveview/live_socket.js [hunk 8]
```diff
@@ -5,12 +5,25 @@ import { LiveSocket } from "phoenix_live_view"
 let csrfToken = document.querySelector("meta[name='csrf-token']")
 let websocketUrl = document.querySelector("meta[name='websocket-url']")
 if (csrfToken && websocketUrl) {
+  let Hooks = {}
+  Hooks.Metrics = {
+    mounted() {
+      this.handleEvent("send_metrics", ({ event_name, params }) => {
+        const afterMetrics = () => {
+          this.pushEvent("send_metrics_after", {event_name, params})
```
Comment (aerosol) [inline assets/js/liveview/live_socket.js:13]: Minor and feel free to disregard -- I think in most (all?) liveviews we concatenate event names with a hyphen.

Comment (zoldar) [reply assets/js/liveview/live_socket.js:13]: Oh, I'll double check and correct it then 👍

Comment (zoldar) [reply assets/js/liveview/live_socket.js]: Done with slightly extended tests in https://github.com/plausible/analytics/pull/3344/commits/74c92520d216361866b2154bd4cd9d43ed55d578. Thanks 🙌

File: lib/plausible_web/router.ex [hunk 9]
```diff
@@ -132,9 +136,27 @@ defmodule PlausibleWeb.Router do
   scope "/", PlausibleWeb do
     pipe_through [:browser, :csrf]
 
-    get "/register", AuthController, :register_form
+    scope alias: Live, assigns: %{connect_live_socket: true} do
+      pipe_through [PlausibleWeb.RequireLoggedOutPlug, :focus_layout]
+
+      scope assigns: %{disable_registration_for: [:invite_only, true]} do
+        pipe_through PlausibleWeb.Plugs.MaybeDisableRegistration
```
Comment (ruslandoga) [inline lib/plausible_web/router.ex:143]: Plugs are not called in liveview-redirects, maybe hooks should be instead? Also, could this logic be moved to the liveview module? It's a bit difficult to find in the router...

**Pull request merged**